src/helpers/network.rs

```diff
@@ -1,101 +1,80 @@
-    protocol::{RecordId, Step},
+    helpers::{
+        transport::{NetworkEventData, SubscriptionType, Transport, TransportCommand},
+        Error, HelperIdentity, Role,
+    },
+    protocol::{QueryId, Step},
 };
-use async_trait::async_trait;
-use futures::{ready, Stream};
-use pin_project::pin_project;
-use std::fmt::{Debug, Formatter};
-use std::pin::Pin;
-use std::task::{Context, Poll};
-use tokio::sync::mpsc;
-use tokio_util::sync::{PollSendError, PollSender};
+use futures::{Stream, StreamExt};
+use std::collections::HashMap;
 
 /// Combination of helper role and step that uniquely identifies a single channel of communication
 /// between two helpers.
-#[derive(Clone, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
 pub struct ChannelId {
     pub role: Role,
     pub step: Step,
 }
 
-#[derive(Debug, PartialEq, Eq)]
-pub struct MessageEnvelope {
-    pub record_id: RecordId,
-    pub payload: MessagePayload,
-}
-
 pub type MessageChunks = (ChannelId, Vec<u8>);
 
-/// Network interface for components that require communication.
-#[async_trait]
-pub trait Network: Sync {
-    /// Type of the channel that is used to send messages to other helpers
-    type Sink: futures::Sink<MessageChunks, Error = Error> + Send + Unpin + 'static;
-    type MessageStream: Stream<Item = MessageChunks> + Send + Unpin + 'static;
-
-    /// Returns a sink that accepts data to be sent to other helper parties.
-    fn sink(&self) -> Self::Sink;
-
-    /// Returns a stream to receive messages that have arrived from other helpers. Note that
-    /// some implementations may panic if this method is called more than once.
-    fn recv_stream(&self) -> Self::MessageStream;
-}
-
-impl ChannelId {
-    #[must_use]
-    pub fn new(role: Role, step: Step) -> Self {
-        Self { role, step }
-    }
-}
-
-impl Debug for ChannelId {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        write!(f, "channel[{:?},{:?}]", self.role, self.step)
-    }
+/// Given any implementation of [`Transport`], a `Network` is able to send and receive
+/// [`MessageChunks`] for a specific query id. The [`Transport`] will receive `NetworkEvents`
+/// containing the `MessageChunks`
+pub struct Network<T> {
+    transport: T,
+    query_id: QueryId,
+    roles_to_helpers: HashMap<Role, HelperIdentity>,
 }
 
-/// Wrapper around a [`PollSender`] to modify the error message to match what the [`NetworkSink`]
-/// requires. The only error that [`PollSender`] will generate is "channel closed", and thus is the
-/// only error message forwarded from this [`NetworkSink`].
-#[pin_project]
-pub struct NetworkSink<T> {
-    #[pin]
-    inner: PollSender<T>,
-}
-
-impl<T: Send + 'static> NetworkSink<T> {
-    #[must_use]
-    pub fn new(sender: mpsc::Sender<T>) -> Self {
+impl<T: Transport> Network<T> {
+    pub fn new(
+        transport: T,
+        query_id: QueryId,
+        roles_to_helpers: HashMap<Role, HelperIdentity>,
```

Comment:
arrays are indexable by `Role` and hashmap lookup has a cost, I think `[HelperIdentity; 3]` is better